### ISSUE TITLE
Select-String -Quiet not working as documented

### ISSUE THREAD

**Issue by FastEddy1114** [PowerShell/PowerShell#16681]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Select-Object -Quiet does not return $false when the -Pattern is not found as documented at https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/select-string?view=powershell-7.2 It actually returns $null instead if the -Pattern is not found.

Simple test is as follows:
`"xxxHELLOaaa" | Select-String "HELLO" -SimpleMatch -Quiet`
returns $true as expected

`"xxxHELLOaaa" | Select-String "HaLLO" -SimpleMatch -Quiet`
returns nothing/$null unexpectedly, it should return $false

### Expected behavior

```console
`"xxxHELLOaaa" | Select-String "HaLLO" -SimpleMatch -Quiet`
$false
```

### Actual behavior

```console
`"xxxHELLOaaa" | Select-String "HaLLO" -SimpleMatch -Quiet`
$null
```

### Environment data

```powershell
$PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![pwsh-screenshot](https://user-images.githubusercontent.com/47258351/147780106-1ae7fc1a-f38b-4c03-a0e3-0489c176cfa5.png)

**Comment by kasini3000** [comment]
I think it’s a bug,
Select-String -Pattern "xxx" -Path'd:\xxx2.txt' -Quiet # can return false

**Comment by FastEddy1114** [comment]
> I think it’s a bug, Select-String -Pattern "xxx" -Path 'd:\xxx2.txt' -Quiet # can return false

You're right searching for a pattern within a file does work and return $true or $false as expected.  It is when piping in data or using a variable as the -InputObject that it fails to return $false if the pattern is not found.

`$Text = "hello"`
`Select-String -InputObject $Text -Pattern "hello" -Quiet` returns $true as expected.
`Select-String -InputObject $Text -Pattern "hallo" -Quiet` returns $null unexpectedly.

**Comment by 237dmitry** [comment]
I think this is not matter as `$null ? $true : $false` returns `False`
Another thing is that the documentation states that if there is a mismatch `False` should be returned.

**Comment by jhoneill** [comment]
> I think this is not matter as `$null ? $true : $false` returns `False` Another thing is that the documentation states that if there is a mismatch `False` should be returned.

It is only a problem with something like ` $matches =  select-string  -Quiet -Pattern "foobar" -InputObject $text`  if the code then checks for `$matches -eq $false`  `if ($matches)` or `if (-not $matches)` are fine.  

The behaviour is odd AND doesn't match what the documentation says  
`dir *.xml | select-string -quiet -pattern "foobar" `  returns true or false **for each file** so we get this .  
```
> (dir *.xml | select-string  -Quiet -Pattern roobaf)
False
False
False
```
Any multi-member array is treated as Boolean true even if every member is false so 
```
> [boolean](dir *.xml | select-string  -Quiet -Pattern roobaf)
True
```

However  `select-string -Quiet -Pattern Bearer -Path *.xml` returns a single false

I can imagine both cases being relied on so any change is likely to be a breaking one. So it looks like the answer is to fix the documentation.

**Comment by kasini3000** [comment]
hi @jhoneill :
Your example brings unnecessary confusion.
Must exclude select-string for multiple elements, because this theme does not consider $true, $false, $true.
Only limited to: select-string -Quiet -Pattern xxx -Path onefile.txt

**Comment by jhoneill** [comment]
@kasini3000  
To reduce the confusion: 

- With piped text the result is `true` or `null`
- With a single file the the result is `true` or `false` 
- With multiple piped files a `true` or `false` _per file_ is returned  

Most tests will treat `null` as false so the inconsistency the OP reported is only an issue if code explicitly checks for `$false`

The behaviour is inconsistent; _if any change is made making the piped files behave the same way as a wild card value for the path parameter should also be considered._ 
```
ps >  select-string -path .\Expand*.ps1 -Pattern excel -Quiet
True  # three files began "expand" and contain "excel" 

ps >  select-string -path .\Expand*.ps1 -Pattern ecxel -Quiet
False  # three files began "expand" and did not contain "ecxel"

ps >  dir .\Expand*.ps1   | select-string -Pattern ecxel -Quiet
False
False
False 
# if file objects are piped in I get an array result which will be treated as true

ps > type .\Expand*.ps1   | select-string -Pattern ecxel -Quiet
#returns  null  because text streams don't return false

ps > type .\Expand*.ps1   | select-string -Pattern excel -Quiet
True  # a single true because the files are one text steam
```

**Comment by kasini3000** [comment]
@jhoneill jhoneill
To reduce the confusion: 
Multiple elements may be return $true, $false, $true. or $true, $true. $false, or more ... 


* only talk Single file. Do you know what is a single file ?
* only talk Single element on the right side of the pipe .eg: 
type .\Expand*.ps1   | foreach-object { select-string -Pattern excel -Quiet -file $_}
* only talk  -Quiet

**Comment by jhoneill** [comment]
@kasini3000   I don't see any need the limit the discussion to only one of the ways that `-Quiet` does not behave as documented.  
>    `-Quiet` Indicates that the cmdlet returns a Boolean value (True or False),  

The OP wasn't passing any files, but piping text in which case it doesn't return a Boolean for 'no match found'.  You said it was a bug, running with `-Path` works: True on both counts:  if the `-Path` specifies one file or many the result is a boolean, as the documentation says.  It is sometimes necessary to pipe file objects which returns an array of Booleans instead of a Boolean.

**Comment by kasini3000** [comment]
@FastEddy1114 
To reduce ambiguity, would you like to change the title to ?

```
Select-String -Quiet return $null instead of $false on single file or element when pipe or -InputObject
```

**Comment by FastEddy1114** [comment]
> @FastEddy1114 To reduce ambiguity, would you like to change the title to ?
> 
> ```
> Select-String -Quiet return $null instead of $false on single file or element when pipe or -InputObject
> ```

No, I really don't think changing the title is necessary.  While yes, I discovered this using piped text and in -InputObject variable array that doesn't stop the inconsistencies that @jhoneill found from being true as well.  The title still makes sense that the command is not working as documented, there are just more of these cases known now.  So either the code needs updated to work as documented in all cases or the documentation needs updated to clarify the expected results based on the use case of the command.
